FEAT: Use platform specific secret generators in `Serializor`

oruborus opened the pull request:
I think this might simplify the secret generation process and relies less on various (potentially editable) files.

Currently 'BSD' and 'Solaris' are not supported. 

I can test this only on windows as I have no access to other machines - so testing on the other platforms is advised.

oruborus committed:
FEAT: Use platform specific secret generators in `Serializor`

frodeborli commented:
I like this contribution, but I want a couple of changes:

For linux, the /etc/machine-id is the first file it should search for. It is the fastest possible way to get a machine id from linux I think. If we can avoid launching a child process via shell_exec, then that is always preferable. For Windows, I think your approach is great. Also, the Linux secret generator most likely will work on Solaris and similar unix based systems I *believe*. Perhaps there could be a fallback secret generator which creates a file in sys_get_temp_dir() with a secret?

oruborus commented:
> For linux, the /etc/machine-id is the first file it should search for. It is the fastest possible way to get a machine id from linux I think. If we can avoid launching a child process via shell_exec, then that is always preferable. 

I think avoiding `shell_exec` for speed would not matter that much as the secret is normally set once per request.
Regarding the ordering of the retrieval: as the 3 values are concatenated in the end, changing the order would only speed up the time to failue...

> Perhaps there could be a fallback secret generator which creates a file in sys_get_temp_dir() with a secret?

Do you mean storing (when non-existant) and retrieving a (pseudo-)random hash-value? I will add that.

> Also, the Linux secret generator most likely will work on Solaris and similar unix based systems I believe.

I don't think so, but I found some potencially viable ways. (testing will be an issue...)

oruborus committed:
TEST: Add missing coverage information

oruborus committed:
FIX: Remove use statement with non-compound name

oruborus committed:
FEAT: Implement `FallbackSecretGenerator`

This gets used when either the platform specific secret generator fails
or no suitable generator was found.

oruborus committed:
FIX: Correct coverage information

oruborus committed:
FEAT: Implement `CanExecuteCommand` trait for use in SecretGenerators

This helper encapsulates `proc_open` and swallows any output made to
STDERR

oruborus committed:
REFACTOR: Make use of `CanExecuteCommand` trait

oruborus committed:
FEAT: Implement secret generators for Solaris and BSD

oruborus committed:
REFACTOR: Add a factory for secret generation

oruborus commented:
@frodeborli What do you think about these changes?

frodeborli commented:
Thanks for the contribution! I merged it into dev and will test it on the platforms I can before tagging a minor version.

oruborus commented:
Great.
Now I can work on improved testing without hacking around the missing secret for my machine.
